src/v/ssx/thread_worker.h

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "utils/mutex.h"
+#include "vassert.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/alien.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/internal/pollable_fd.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/std-coroutine.hh>
+
+#include <sys/eventfd.h>
+
+#include <exception>
+#include <memory>
+#include <thread>
+
+namespace ssx {
+
+namespace impl {
+
+class task_base {
+public:
+    task_base() = default;
+    task_base(task_base&&) = delete;
+    task_base(task_base const&) = delete;
+    task_base& operator=(task_base&&) = delete;
+    task_base& operator=(task_base const&) = delete;
+
+    virtual void process(ss::alien::instance&, ss::shard_id) = 0;
+    virtual void
+    set_exception(ss::alien::instance&, ss::shard_id, std::exception_ptr)
+      = 0;
+    virtual ~task_base() = default;
+};
+
+template<typename Func>
+class worker_task final : public task_base {
+    using futurator = ss::futurize<std::invoke_result_t<Func>>;
+
+public:
+    explicit worker_task(Func func)
+      : _func{std::move(func)} {}
+
+    typename futurator::type get_future() noexcept {
+        return _promise.get_future();
+    }
+
+    void process(ss::alien::instance& alien, ss::shard_id shard) final {
+        auto f = futurator::invoke(_func);
+        ss::alien::run_on(alien, shard, [this, f{std::move(f)}]() mutable {
+            f.forward_to(std::move(_promise));
+        });
+    }
+
+    void set_exception(
+      ss::alien::instance& alien,
+      ss::shard_id shard,
+      std::exception_ptr p) final {
+        ss::alien::run_on(
+          alien, shard, [this, p]() mutable { _promise.set_exception(p); });
+    }
+
+private:
+    Func _func;
+    typename futurator::promise_type _promise;
+};
+
+class thread_worker {
+    static constexpr eventfd_t RUNNABLE_INIT = 0;
+    static constexpr eventfd_t RUNNABLE_READY = 1;
+    static constexpr eventfd_t RUNNABLE_STOP = 2;
+
+public:
+    thread_worker()
+      : _alien{ss::engine().alien()}
+      , _shard_id(ss::this_shard_id()) {}
+
+    void start() {
+        _worker = std::thread([this]() { return run(); });
+    }
+
+    ss::future<> stop() {
+        co_await _gate.close();
+        _ready.signal(RUNNABLE_STOP);
+        if (_worker.joinable()) {
+            _worker.join();
```

Comment:
I guess I could be more explicit here - it's certainly designed for start and stop to be tied to the application lifetime.

The idea is to async wait for all operations to finish with `gate::wait`, so there shouldn't be any significant work left on the thread at the time it's joined, just destruction.

Reply:
>significant work left on the thread at the time it's joined, just destruction.

join waits for the thread to exit, so i'd assume it might block and cause a reactor stall. i don't know if that is true, and that's not a problem since it's tied to app lifetime, but something i would expect we write a comment about to warn users.